File: test/os_time.c
/* Behavioral contract for the host clock services, run against whichever
 * implementation the build selected (`-Dos-time=c` or the Zig default).
 *
 * Clocks cannot be pinned to fixed vectors the way a pure kernel can, so this
 * contract fixes their invariants instead: ranges, orderings, the fallback for
 * an unrecognized source, and that a sleep actually advances a monotonic
 * clock. */

#include <assert.h>
#include <stdint.h>
#include <stdio.h>
#include <time.h>
#include <janet.h>

/* Declared rather than included from src/core/util.h, so the contract depends
 * only on the internal ABI it exercises. */
enum JanetTimeSource {
    JANET_TIME_REALTIME,
    JANET_TIME_MONOTONIC,
    JANET_TIME_CPUTIME
};
int janet_gettime(struct timespec *spec, enum JanetTimeSource source);
double janet_os_time_now(void);
void janet_os_sleep(double seconds);

/* Any run of this test is after the start of 2023 and before the end of 2200. */
#define EPOCH_LOWER_BOUND 1672531200LL
#define EPOCH_UPPER_BOUND 7289654400LL

static double seconds_of(struct timespec spec) {
    return (double) spec.tv_sec + (double) spec.tv_nsec / 1e9;
}

static void check_normalized(struct timespec spec) {
    assert(spec.tv_nsec >= 0);
    assert(spec.tv_nsec < 1000000000L);
}

static void test_realtime(void) {
    struct timespec spec;
    assert(janet_gettime(&spec, JANET_TIME_REALTIME) == 0);
    check_normalized(spec);
    assert((int64_t) spec.tv_sec > EPOCH_LOWER_BOUND);
    assert((int64_t) spec.tv_sec < EPOCH_UPPER_BOUND);

    /* The wall clock behind os/time agrees with the real-time source. */
    double now = janet_os_time_now();
    assert(now > (double) EPOCH_LOWER_BOUND);
    assert(now < (double) EPOCH_UPPER_BOUND);
    assert(now - (double) spec.tv_sec >= -2.0);
    assert(now - (double) spec.tv_sec <= 2.0);
}

static void test_monotonic(void) {
    struct timespec first, second;
    assert(janet_gettime(&first, JANET_TIME_MONOTONIC) == 0);
    assert(janet_gettime(&second, JANET_TIME_MONOTONIC) == 0);
    check_normalized(first);
    check_normalized(second);
    assert(seconds_of(second) >= seconds_of(first));
}

static void test_cputime(void) {
    struct timespec before, after;
    volatile double sink = 0;
    assert(janet_gettime(&before, JANET_TIME_CPUTIME) == 0);
    check_normalized(before);
    for (int32_t i = 0; i < 8000000; i++) sink += (double) i;
    assert(janet_gettime(&after, JANET_TIME_CPUTIME) == 0);
    check_normalized(after);
    /* Consumed CPU time is measured from process start, so it is positive and
     * never decreases. */
    assert(seconds_of(before) > 0);
    assert(seconds_of(after) >= seconds_of(before));
}

static void test_unknown_source(void) {
    struct timespec realtime, unknown;
    assert(janet_gettime(&realtime, JANET_TIME_REALTIME) == 0);
    /* An unrecognized source falls back to the real-time clock rather than
     * failing, because the C shim initializes its clock id before testing. */
    assert(janet_gettime(&unknown, (enum JanetTimeSource) 99) == 0);
    check_normalized(unknown);
    assert(seconds_of(unknown) - seconds_of(realtime) >= -2.0);
    assert(seconds_of(unknown) - seconds_of(realtime) <= 2.0);
}

static void test_sleep(void) {
    struct timespec before, after;
    double elapsed;

    assert(janet_gettime(&before, JANET_TIME_MONOTONIC) == 0);
    janet_os_sleep(0.05);
    assert(janet_gettime(&after, JANET_TIME_MONOTONIC) == 0);
    elapsed = seconds_of(after) - seconds_of(before);
    /* The lower bound is the contract; the upper bound is loose enough to
     * survive a loaded machine while still catching a sleep that multiplied
     * its argument by the wrong factor. */
    assert(elapsed >= 0.04);
    assert(elapsed < 10.0);

    /* A zero delay returns promptly rather than blocking. */
    assert(janet_gettime(&before, JANET_TIME_MONOTONIC) == 0);
    janet_os_sleep(0);
    assert(janet_gettime(&after, JANET_TIME_MONOTONIC) == 0);
    assert(seconds_of(after) - seconds_of(before) < 10.0);
}

static void run(JanetTable *env, const char *source) {
    Janet result;
    assert(janet_dostring(env, source, "os-time-contract", &result) == 0);
}

static void test_core_functions(void) {
    JanetTable *env = janet_core_env(NULL);

    /* Every documented source and format. */
    run(env,
        "(assert (number? (os/clock)))\n"
        "(assert (number? (os/clock :realtime)))\n"
        "(assert (number? (os/clock :monotonic)))\n"
        "(assert (number? (os/clock :cputime)))\n"
        "(assert (number? (os/clock :realtime :double)))\n");

    run(env,
        "(def whole (os/clock :realtime :int))\n"
        "(assert (= whole (math/floor whole)))\n"
        "(def parts (os/clock :monotonic :tuple))\n"
        "(assert (tuple? parts))\n"
        "(assert (= 2 (length parts)))\n"
        "(assert (= (parts 0) (math/floor (parts 0))))\n"
        "(assert (>= (parts 1) 0))\n"
        "(assert (< (parts 1) 1000000000))\n");

    /* The real-time clock and os/time agree; the monotonic clock does not go
     * backwards. */
    run(env,
        "(assert (< (math/abs (- (os/time) (os/clock :realtime :int))) 2))\n"
        "(assert (> (os/time) 1672531200))\n"
        "(def earlier (os/clock :monotonic))\n"
        "(assert (>= (os/clock :monotonic) earlier))\n");

    /* Sleeping advances the monotonic clock. */
    run(env,
        "(def before (os/clock :monotonic))\n"
        "(os/sleep 0.05)\n"
        "(def elapsed (- (os/clock :monotonic) before))\n"
        "(assert (>= elapsed 0.04))\n"
        "(assert (< elapsed 10))\n"
        "(assert (nil? (os/sleep 0)))\n");

    /* Invalid arguments keep raising from the C validation boundary. */
    run(env,
        "(assert (not (first (protect (os/clock :nope)))))\n"
        "(assert (not (first (protect (os/clock :realtime :nope)))))\n"
        "(assert (not (first (protect (os/sleep -1)))))\n");
}

int main(void) {
    test_realtime();
    test_monotonic();
    test_cputime();
    test_unknown_source();
    test_sleep();

    janet_init();
    test_core_functions();
    janet_deinit();
    return 0;
}

File: test/os_time.zig
//! Behavioral contract for the host clock services: `os.gettimeAbi`'s three
//! sources, the wall clock behind `os/time`, and `os.sleepFor`.
//!
//! A clock cannot be pinned to fixed vectors the way a parser can, so what is
//! asserted here is a set of *invariants* rather than values: which ranges each
//! source falls in, which orderings two readings guarantee, what an
//! unrecognised source does, and that a sleep actually advances a monotonic
//! clock. Those are the properties a port can break while still returning
//! plausible numbers, and no Janet suite reaches `os.gettimeAbi` at all.
//!
//! ## The bounds are deliberately wide
//!
//! Every numeric bound below is loose on purpose, and the reason is that this
//! runs on a shared machine under a matrix at `-j2`. A tight upper bound on a
//! sleep would fail for scheduling reasons and cost an afternoon proving it.
//! So the *lower* bound is the contract and the upper bound is only there to
//! catch a sleep that multiplied its argument by the wrong factor.
//!
//! ## The fallback for an unknown source is established behaviour
//!
//! `os.gettimeAbi` given a source it does not recognise gives the real-time
//! clock and reports success rather than failing, because it initialises its
//! clock id before it tests the source. That is asserted here so a change to
//! it would be deliberate.

// ==========================================================================
// Standard library imports
// ==========================================================================

const std = @import("std");

// ==========================================================================
// Project imports
// ==========================================================================

const core_env = @import("subsystems").env;
const expect = @import("expect.zig").expect;
const harness = @import("harness.zig");
const os = @import("subsystems").os;
const repr = @import("repr");
const tables = @import("subsystems").value.tables;
const value = @import("subsystems").value;
const vm_lifecycle = @import("subsystems").lifecycle;
const wrap = @import("subsystems").value.wrap;

// ==========================================================================
// Constants
// ==========================================================================

var environment: *tables.Table = undefined;

/// Any run of this is after the start of 2023 and before the end of 2200.
const epoch_lower = 1672531200;
const epoch_upper = 7289654400;

// ==========================================================================
// Types
// ==========================================================================

/// The three clock sources, restated here so that a renumbering on either side
/// fails rather than agreeing with itself.
const Source = enum(c_int) {
    realtime = 0,
    monotonic = 1,
    cputime = 2,
    /// Not a source, and the value that reaches the fallback below.
    unrecognised = 99,
};

/// The `struct timespec` the entry point fills, declared here rather than
/// borrowed from `os.zig`: the field offsets are what this contract is
/// checking, so reading them from the subject would make the check circular.
const TimeSpec = extern struct {
    seconds: isize,
    nanoseconds: isize,
};

// ==========================================================================
// Cases
// ==========================================================================

fn read(source: Source) TimeSpec {
    // `os.Timespec` is the host's `struct timespec` and this is the contract's
    // independent restatement of it, so the two are separate types with the
    // same fields. On riscv32 the host's is more strictly aligned than a pair
    // of `isize`s, which is what the `@alignCast` says, and which is itself
    // a claim: if the two ever disagreed on size or offsets, the fields read
    // below would be wrong.
    var spec: TimeSpec align(@alignOf(os.Timespec)) = undefined;
    expect(os.gettimeAbi(@ptrCast(&spec), @bitCast(@intFromEnum(source))) == 0);
    // Normalised: the nanosecond field is a remainder, not a free-running
    // count, so a nanosecond field at or above a second would show up here.
    expect(spec.nanoseconds >= 0);
    expect(spec.nanoseconds < 1_000_000_000);
    return spec;
}

fn seconds(spec: TimeSpec) f64 {
    return @as(f64, @floatFromInt(spec.seconds)) +
        @as(f64, @floatFromInt(spec.nanoseconds)) / 1e9;
}

fn eval(source: [*:0]const u8) void {
    var result: repr.Value = undefined;
    expect(core_env.dostring(environment, source, "os-time-contract", &result) == 0);
}

fn theRealtimeClock() void {
    const spec = read(.realtime);
    expect(spec.seconds > epoch_lower);
    expect(spec.seconds < epoch_upper);

    // The wall clock behind `os/time` is the same clock, so the two agree to
    // within the time it takes to call them twice.
    const now = os.timeNow();
    expect(now > @as(f64, epoch_lower));
    expect(now < @as(f64, epoch_upper));
    expect(@abs(now - @as(f64, @floatFromInt(spec.seconds))) <= 2.0);
}

fn theMonotonicClockDoesNotGoBackwards() void {
    const first = read(.monotonic);
    const second = read(.monotonic);
    expect(seconds(second) >= seconds(first));
}

fn theCputimeClockAccumulates() void {
    const before = read(.cputime);

    // Work the optimiser cannot remove, so that the clock has something to
    // measure: a mutable sink, read afterwards so nothing can fold it away.
    var sink: f64 = 0;
    var i: i32 = 0;
    while (i < 8_000_000) : (i += 1) sink += @floatFromInt(i);
    expect(sink > 0);

    const after = read(.cputime);
    // Measured from process start, so it is positive and never decreases.
    expect(seconds(before) > 0);
    expect(seconds(after) >= seconds(before));
}

/// An unrecognised source falls back to the real-time clock; see the header.
fn anUnknownSourceIsTheRealtimeClock() void {
    const realtime = read(.realtime);
    const unknown = read(.unrecognised);
    expect(@abs(seconds(unknown) - seconds(realtime)) <= 2.0);
}

fn sleepingAdvancesTheMonotonicClock() void {
    const before = read(.monotonic);
    os.sleepFor(0.05);
    const after = read(.monotonic);

    const elapsed = seconds(after) - seconds(before);
    // The lower bound is the contract; see the header on the upper one.
    expect(elapsed >= 0.04);
    expect(elapsed < 10.0);

    // A zero delay returns rather than blocking.
    const idle_before = read(.monotonic);
    os.sleepFor(0);
    const idle_after = read(.monotonic);
    expect(seconds(idle_after) - seconds(idle_before) < 10.0);
}

/// `os/clock` takes a source and a format, and the combinations are what the
/// suites do not cover. Written in Janet rather than in Zig because each of
/// these is one assertion about a returned value's shape, which Janet says in
/// a quarter of the space, and because `:tuple` returns a tuple of two
/// numbers whose second field is a nanosecond remainder, which is far easier
/// to state as a predicate than to unwrap.
fn theSourcesAndFormats() void {
    // Absent from a reduced-OS build, and `options.os_time` does not say so:
    // that field is `hasGettime`, which is true there because the *subsystem*
    // is still compiled, the event loop needing `os.gettimeAbi` whether or not
    // `os/clock` is registered. The kernels above run either way; only this
    // section and `theRefusals` depend on the registration.
    if (harness.coreOptional("os/clock") == null) return;

    eval(
        \\(assert (number? (os/clock)))
        \\(assert (number? (os/clock :realtime)))
        \\(assert (number? (os/clock :monotonic)))
        \\(assert (number? (os/clock :cputime)))
        \\(assert (number? (os/clock :realtime :double)))
    );

    eval(
        \\(def whole (os/clock :realtime :int))
        \\(assert (= whole (math/floor whole)))
        \\(def parts (os/clock :monotonic :tuple))
        \\(assert (tuple? parts))
        \\(assert (= 2 (length parts)))
        \\(assert (= (parts 0) (math/floor (parts 0))))
        \\(assert (>= (parts 1) 0))
        \\(assert (< (parts 1) 1000000000))
    );

    eval(
        \\(assert (< (math/abs (- (os/time) (os/clock :realtime :int))) 2))
        \\(assert (> (os/time) 1672531200))
        \\(def earlier (os/clock :monotonic))
        \\(assert (>= (os/clock :monotonic) earlier))
    );

    eval(
        \\(def before (os/clock :monotonic))
        \\(os/sleep 0.05)
        \\(def elapsed (- (os/clock :monotonic) before))
        \\(assert (>= elapsed 0.04))
        \\(assert (< elapsed 10))
        \\(assert (nil? (os/sleep 0)))
    );
}

/// The refusals, which validation makes above the kernels. Each is read as a
/// value rather than through `protect`.
fn theRefusals() void {
    const clock = harness.coreOptional("os/clock") orelse return;
    const sleep = harness.coreOptional("os/sleep") orelse return;
    var argument: [2]repr.Value = undefined;

    argument[0] = value.fromBytes("nope", .keyword);
    expect(harness.raised(clock, .{argument[0..1]}) != null);

    argument[0] = value.fromBytes("realtime", .keyword);
    argument[1] = value.fromBytes("nope", .keyword);
    expect(harness.raised(clock, .{argument[0..2]}) != null);

    // A negative sleep is refused rather than treated as zero.
    argument[0] = harness.wrapInteger(-1);
    expect(harness.raised(sleep, .{argument[0..1]}) != null);

    // So is a NaN, which names no duration.
    argument[0] = wrap.fromNumber(std.math.nan(f64));
    expect(harness.raised(sleep, .{argument[0..1]}).?.says("invalid argument to sleep"));
}

// ==========================================================================
// Entry
// ==========================================================================

pub fn run() void {
    theRealtimeClock();
    theMonotonicClockDoesNotGoBackwards();
    theCputimeClockAccumulates();
    anUnknownSourceIsTheRealtimeClock();
    sleepingAdvancesTheMonotonicClock();

    harness.init();
    environment = harness.coreEnv();
    theSourcesAndFormats();
    theRefusals();
    vm_lifecycle.deinit();
}
